src/diff_edit.rs

```diff
@@ -168,89 +181,143 @@ pub fn run_mergetool(
                 repo_path.clone(),
                 summary,
             ));
         }
     };
     if content.removes.len() > 1 || content.adds.len() > 2 {
-        return Err(DiffEditError::ConflictTooComplicatedError {
+        return Err(ConflictResolveError::ConflictTooComplicatedError {
             path: repo_path.clone(),
             removes: content.removes.len(),
             adds: content.adds.len(),
         });
     };
 
+    let editor = get_merge_tool_from_settings(ui, settings)?;
     let temp_dir = tempfile::Builder::new()
         .prefix("jj-merge-")
         .tempdir()
-        .map_err(DiffEditError::SetUpDirError)?;
+        .map_err(ExternalToolError::SetUpDirError)?;
     let files_maybe = [
         ("base", content.removes.get(0)),
         ("left", content.adds.get(0)),
         ("right", content.adds.get(1)),
-        ("output", Some(&materialized_conflict)),
+        (
+            "output",
+            if editor.merge_tool_edits_conflict_markers {
+                Some(&materialized_conflict)
+            } else {
+                None
+            },
+        ),
     ];
     let files: HashMap<&str, _> = files_maybe
         .map(|(name, f)| (name, f.cloned().unwrap_or_default()))
         .iter()
         .cloned()
         .collect();
-    let paths: Result<HashMap<&str, _>, DiffEditError> = files
+    let paths: Result<HashMap<&str, _>, ConflictResolveError> = files
         .iter()
         .map(|(name, contents)| {
             let path = temp_dir.path().join(name);
-            let mut file = File::create(&path).map_err(DiffEditError::SetUpDirError)?;
+            let mut file = File::create(&path).map_err(ExternalToolError::SetUpDirError)?;
             file.write_all(contents)
-                .map_err(DiffEditError::SetUpDirError)?;
-            file.sync_all()?;
+                .map_err(ExternalToolError::SetUpDirError)?;
+            file.sync_all().map_err(ExternalToolError::SetUpDirError)?;
             drop(file);
             if *name != "output" {
                 // TODO: Should actually ignore the error here, or have a warning.
-                set_readonly_recursively(&path)?;
+                set_readonly_recursively(&path).map_err(ExternalToolError::SetUpDirError)?;
             }
             Ok((*name, path))
         })
         .collect();
     let paths = paths?;
 
-    let progname = "vimdiff";
-    let exit_status = Command::new(progname)
-        .args(["-f", "-d"])
-        .arg(paths.get("output").unwrap())
-        .arg("-M")
-        .args(["left", "base", "right"].map(|n| paths.get(n).unwrap()))
-        .args(["-c", "wincmd J", "-c", "setl modifiable write"])
+    let args = interpolate_mergetool_filename_patterns(&editor.merge_args, &paths);
+    let args_str = args
+        .iter()
+        .map(|p| {
+            p.to_str()
+                .unwrap_or("<path with invalid unicode>")
+                .to_string()
```

Comment:
Probably better to use `to_string_lossy().into_owned()` instead since this is just for an error message.

Alternatively, `interpolate_mergetool_filename_patterns()` could enforce the command arguments are valid unicode `String`s. `jj` doesn't support non-utf-8 paths anyway.